counsel.el

```diff
@@ -4517,6 +4517,82 @@ If there is no such buffer, start a new `shell' with NAME."
                             (reusable-frames . visible)))
     (shell name)))
 
+;;** `counsel-firefox-bookmarks'
+(ivy-set-actions
+ 'counsel-firefox-bookmarks
+ `(("n" ,(lambda (x) (kill-new (second x))) "copy name")
+   ("l" ,(lambda (x) (kill-new (third x))) "copy location")))
+
+(defvar counsel-firefox-bookmarks-file
+  (car (file-expand-wildcards "~/.mozilla/firefox/*/bookmarks.html"))
+  "Firefox's automatically exported HTML bookmarks file.")
+
+(defface counsel-firefox-bookmarks-tag
+  '((t :inherit font-lock-comment-face))
+  "Face used by `counsel-firefox-bookmarks' for tags."
+  :group 'ivy-faces)
+
+(defface counsel-firefox-bookmarks-location
+  '((t :inherit link))
+  "Face used by `counsel-firefox-bookmarks' for locations."
+  :group 'ivy-faces)
+
+(defun counsel-firefox-bookmarks-action (x)
+  "Browse candidate X."
+  (browse-url (third x)))
+
+(declare-function xml-substitute-special "xml")
+
+(defun counsel--firefox-bookmarks-candidates ()
+  "Return list of `counsel-firefox-bookmarks' candidates."
+  (unless (and counsel-firefox-bookmarks-file
+               (file-readable-p counsel-firefox-bookmarks-file))
+    (signal 'file-error (list "Opening `counsel-firefox-bookmarks-file'"
+                              "No such readable file"
+                              counsel-firefox-bookmarks-file)))
+  (require 'xml)
+  (with-temp-buffer
+    (insert-file-contents counsel-firefox-bookmarks-file)
+    (let ((case-fold-search t)
+          candidates)
+      (while (re-search-forward
+              "<a href=\"\\([^\"]+?\\)\"[^>]*?>\\([^<]*?\\)</a>" nil t)
+        (let* ((a (match-string 0))
+               (href (match-string 1))
+               (name (save-match-data
+                       (xml-substitute-special (match-string 2))))
+               (tags (and (string-match "tags=\"\\([^\"]+?\\)\"" a)
+                          (mapconcat
+                           (lambda (tag)
+                             (put-text-property 0 (length tag) 'face
+                                                'counsel-firefox-bookmarks-tag
+                                                tag)
+                             tag)
+                           (split-string (match-string 1 a) "," t)
+                           ":"))))
+          (push (list
+                 (concat (if tags (concat name " :" tags ":") name)
+                         " "
+                         (propertize href 'face 'counsel-firefox-bookmarks-location))
```

Comment:
This unconditionally appends the `href` even when it is equal to the `name`. Wouldn't it be better if only one of the two copies was presented to the user?

Reply:
done